[PyTorch] use borrow_from_optional_tensor in CPU matmul

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #59062
* #58935

Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

swolchok committed:
[PyTorch] use borrow_from_optional_tensor in CPU matmul

Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

ailzhang reviewed (approved):
Do we want to also document this API somewhere to promote usage of borrow_from_optional_tensor? :D

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] use borrow_from_optional_tensor in CPU matmul"


Saves a reference count bump when the `optional<Tensor>` in question `has_value()`.

Differential Revision: [D28687559](https://our.internmc.facebook.com/intern/diff/D28687559/)

[ghstack-poisoned]
